compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/loops/ProgressionHandlers.kt

```diff
@@ -731,23 +731,26 @@ internal class WithIndexHandler(context: CommonBackendContext, private val visit
     }
 }
 
-/** Builds a [HeaderInfo] for iterables not handled by more specialized handlers. */
-internal class DefaultIterableHandler(private val context: CommonBackendContext) : ExpressionHandler {
+/** Builds a [HeaderInfo] for Iterables not handled by more specialized handlers. */
+internal open class DefaultIterableHandler(private val context: CommonBackendContext) : ExpressionHandler {
 
-    override fun matchIterable(expression: IrExpression) = true
+    protected open val iterableClassSymbol = context.ir.symbols.iterable
+
+    override fun matchIterable(expression: IrExpression) = expression.type.isSubtypeOfClass(iterableClassSymbol)
 
     override fun build(expression: IrExpression, scopeOwner: IrSymbol): HeaderInfo? =
         with(context.createIrBuilder(scopeOwner, expression.startOffset, expression.endOffset)) {
-            val iterableClass = expression.type.getClass()!!
-            val iterator =
-                irCall(iterableClass.functions.single {
-                    it.name == OperatorNameConventions.ITERATOR &&
-                            it.valueParameters.isEmpty()
-                }).apply {
-                    dispatchReceiver = expression
-                }
+            // The lowering operates on subtypes of Iterable. Therefore, the IrType could be
+            // a type parameter bounded by Iterable. When that is the case, we cannot get
+            // the class from the type and instead uses the Iterable.iterator() function.
+            val iteratorFun = expression.type.getClass()?.functions?.single {
+                it.name == OperatorNameConventions.ITERATOR &&
+                        it.valueParameters.isEmpty()
```

Comment:
Shouldn't we also check that there's no extension receiver, or is it not relevant in this case?

Reply:
Member function should always "win out" over any extension function, and there's always an `iterator()` member function on `Iterable`s.

Reply:
I meant that a subclass of `Iterable` can have its own member extension iterator, which could lead to `single` throwing an exception here. Something like this:
```
class MyIterable<T> : Iterable<T> {
    ....
    override fun iterator(): Iterator<T> = ...

    // Not an iterator, although has the name "iterator" and no value parameters!
    fun SomeUnrelatedType.iterator(): Double = ...
}
```

Reply:
@punzki I've been investigating a related problem in our project bootstrap with JVM IR, and fixed it along with this one here, please take a look: https://upsource.jetbrains.com/kotlin/review/KOTLIN-CR-3586